### PR TITLE
 Basic TensorFlow implementation of EM, and associated changes.

### PR DESCRIPTION
**Pull request by andychu**
It can be selected by passing --em-executable as fast_em.sh (wrapper for
Python) instead of the binary compiled from fast_em.cc.

See comments at the top of fast_em.py.

- association_test.R: test all three implementations
- bin/test.sh
  - Add demo for TensorFlow
  - Add demo of early convergence
  - Put the output from each implementation in its own directory.
- fast_em.R: Test exit code when shelling out to --em-executable
- Write number of EM iterations in the assoc-metrics.json output (all
  implementations).  This changes the protocol between the R driver and
  EM implementation.
- Clean up the log output from fast_em.cc

### CODE REVIEW

File: analysis/tensorflow/fast_em.py [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+  Returns:
+    pij: numpy.ndarray (e.g. vector of length 8)
+  """
+  # Placeholder for the value from the previous iteration.
+  pij_in = tf.placeholder(tf.float64, shape=(entry_size,))
+
+  # split along dimension 0
+  v_split = tf.split(0, num_entries, v)
+
+  z_numerator = [report * pij_in for report in v_split]
```
Comment (tkaitchuck) [inline analysis/tensorflow/fast_em.py]: Break this out into a function for clarity.

File: analysis/tensorflow/fast_em.py [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+    max_em_iters: maximum number of EM iterations
+
+  Returns:
+    pij: numpy.ndarray (e.g. vector of length 8)
+  """
+  # Placeholder for the value from the previous iteration.
+  pij_in = tf.placeholder(tf.float64, shape=(entry_size,))
+
+  # split along dimension 0
+  v_split = tf.split(0, num_entries, v)
```
Comment (tkaitchuck) [inline analysis/tensorflow/fast_em.py]: See if there is a better way to do this. Either through a different OP as Olfar suggested or by reusing a constant v_split.

File: analysis/tensorflow/fast_em.sh [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+set -o errexit
+
+readonly THIS_DIR=$(dirname $0)
+
+fast-em() {
+  export LD_LIBRARY_PATH=/usr/local/cuda/lib64
+  export CUDA_HOME=/usr/local/cuda-7.0
+
+  # Never returns
+  exec $THIS_DIR/fast_em.py "$@"
```
Comment (tkaitchuck) [inline analysis/tensorflow/fast_em.sh]: You shouldn't clobber the existing variable for LD_LIBRARY_PATH rather you should append that to it. 
Also you don't need to export, you can just assign the variables on the same line as the exec and the child will inherent them without messing with them in the user's shell afterwards.

**Comment by tkaitchuck** [comment]
Aside from comments above, looks good. Merge when ready.

**Comment by andychu** [comment]
OK, I addressed all the comments, except I put a TODO for the v_split thing.  I agree that is probably the issue, but I'm not sure how to fix it yet and I have to do some other stuff.  Right now it is over 200x slower!

**Pull request merged**